### PR TITLE
Better header handling

### PR DESCRIPTION
**Pull request by Ak-Army**
Just a bit update for a better header handling in curl provider

Sorry, i never create a pull request in github :)

### CODE REVIEW

File: Library/Phalcon/Http/Client/Provider/Curl.php [hunk 1]
```diff
@@ -61,14 +64,14 @@ private function initOptions()
         $this->setOptions(array(
             CURLOPT_RETURNTRANSFER  => true,
             CURLOPT_AUTOREFERER     => true,
-            CURLOPT_FOLLOWLOCATION  => true,
+            //CURLOPT_FOLLOWLOCATION  => true,
```
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: Why?

Comment (Ak-Army) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: Cause there is a lot of problem with it:
CURLOPT_FOLLOWLOCATION cannot be activated when safe_mode is enabled or an open_basedir is set in

So i think, if somebody want to use it, he know how they can turn it on,..

**Comment by sergeyklay** [comment]
Okay. Now can you please provide a small test that covers your changes?

File: Library/Phalcon/Http/Client/Provider/Curl.php [hunk 2]
```diff
@@ -112,21 +118,35 @@ private function send($customHeader = array(), $fullResponse = false)
...
+     *
+     * @param  resource $curl
+     * @param  string   $header
+     * @return int
+     */
+    public function readHeader($curl, $header)
+    {
+        $this->responseHeader .= $header;
+
+        return strlen($header);
```
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: mb `getHeaderLen` ?

File: Library/Phalcon/Http/Client/Provider/Curl.php [hunk 3]
```diff
@@ -103,8 +106,11 @@ private function send($customHeader = array(), $fullResponse = false)
             }
             $header[] = 'Expect:';
         }
-        
+
         $this->setOption(CURLOPT_HTTPHEADER, $header);
+        $this->setOption(CURLOPT_HEADER, false);
+        $this->setOption(CURLOPT_HEADERFUNCTION, array($this, "readHeader"));
```
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: use `[ ]` instead `array( )` please. We are in the 1st century :)

File: Library/Phalcon/Http/Client/Provider/Curl.php [hunk 4]
```diff
@@ -112,21 +118,35 @@ private function send($customHeader = array(), $fullResponse = false)
             throw new HttpException(curl_error($this->handle), $errno);
         }
 
-        $headerSize = curl_getinfo($this->handle, CURLINFO_HEADER_SIZE);
+        //$headerSize = curl_getinfo($this->handle, CURLINFO_HEADER_SIZE);
```
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: Why? If this really not needed, remove it. Don not  leave dirty code (e.g. commented code blocks ) here

File: Library/Phalcon/Http/Client/Provider/Curl.php [hunk 5]
```diff
@@ -181,7 +238,7 @@ public function get($uri, $params = array(), $customHeader = array(), $fullRespo
         $this->setOptions(array(
             CURLOPT_URL           => $uri->build(),
             CURLOPT_HTTPGET       => true,
-            CURLOPT_CUSTOMREQUEST => Method::GET,
+            CURLOPT_CUSTOMREQUEST => 'GET',
```
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: Why?

File: Library/Phalcon/Http/Client/Provider/Curl.php (hunk 2 above)
Comment (Ak-Army) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: Cant be protected cause it callad by the curl.

CURLOPT_HEADERFUNCTION 	A callback accepting two parameters. The first is the cURL resource, the second is a string with the header data to be written. The header data must be written by this callback. Return the number of bytes written.

**Comment by Ak-Army** [comment]
What do u think, how can we test curl options :)

**Comment by sergeyklay** [comment]
easy

File: Library/Phalcon/Http/Client/Provider/Curl.php [hunk 6]
```diff
@@ -61,14 +65,12 @@ private function initOptions()
         $this->setOptions(array(
             CURLOPT_RETURNTRANSFER  => true,
             CURLOPT_AUTOREFERER     => true,
-            CURLOPT_FOLLOWLOCATION  => true,
             CURLOPT_MAXREDIRS       => 20,
-            CURLOPT_HEADER          => true,
             CURLOPT_PROTOCOLS       => CURLPROTO_HTTP | CURLPROTO_HTTPS,
             CURLOPT_REDIR_PROTOCOLS => CURLPROTO_HTTP | CURLPROTO_HTTPS,
-            CURLOPT_USERAGENT       => 'Phalcon HTTP/' . self::VERSION . ' (Curl)',
+            CURLOPT_USERAGENT       => 'Phalcon HTTP/'.self::VERSION.' (Curl)',
```
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: rollback this please

File: Library/Phalcon/Http/Client/Provider/Curl.php (hunk 3 above)
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: what about this

File: Library/Phalcon/Http/Client/Provider/Curl.php [hunk 7]
```diff
@@ -154,17 +172,54 @@ private function initPostFields($params, $useEncoding = true)
...
             CURLOPT_PROXY     => $host,
-            CURLOPT_PROXYPORT => $port
+            CURLOPT_PROXYPORT => $port,
         ));
 
         if (!empty($user) && is_string($user)) {
             $pair = $user;
             if (!empty($pass) && is_string($pass)) {
-                $pair .= ':' . $pass;
+                $pair .= ':'.$pass;
```
Comment (sergeyklay) [inline Library/Phalcon/Http/Client/Provider/Curl.php]: why?

**Pull request closed without merging**